[Dependency Cache] Enable Dependency Cache on CI

ParaskP7 opened the pull request:
Project Thread: paaHJt-7CE-p2
Related: #21550 + #21540
Depends On: [a8c-ci-toolkit#135](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/135) -> [a8c-ci-toolkit#138](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/138)
Required By: [buildkite-ci#578](https://github.com/Automattic/buildkite-ci/pull/578)

---

## Description

This PR enables dependency cache on CI for this project. 💾

The dependency cache mechanism is split between:
- A weekly scheduled **save cache** build (see [commit](1a71683264e2e9db49b9e03046a37cb552910c84) and [buildkite-ci#578](https://github.com/Automattic/buildkite-ci/pull/578) PR), which targets the below tasks only:
  - Mobile App (`assembleJetpackJalapenoDebug`)
  - Lint (`lintJetpackJalapenoDebug`)
  - Unit Tests (`assembleJetpackJalapenoDebugUnitTest assembleDebugUnitTest testClasses`)
  - Android tests (`assembleJetpackJalapenoDebugAndroidTest`)
- A **restore cache** step that is run on targeted [pipeline.yml](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.buildkite/pipeline.yml) related jobs, via their corresponding individual `.sh` scripts:
  - Jetpack App (`prototype-build.sh`)
  - WordPress App (`prototype-build.sh`)
  - Jetpack Lint (`lint.sh`)
  - WordPress Lint (`lint.sh`)
  - WordPress Unit Tests (`run-unit-tests.sh`)
  - Processors Unit Tests (`run-unit-tests.sh`)
  - Image Editor Unit Tests (`run-unit-tests.sh`)
  - FluxC Unit Tests (`run-unit-tests.sh`)
  - Login Unit Tests (`run-unit-tests.sh`)
  - [Disabled**] Jetpack Android Tests (`run-instrumented-tests.sh`)
  - [Disabled**] WordPress Android Tests (`run-instrumented-tests.sh`)

PS.1: The `detekt` and `checkstyle` job aren't targeted because it only takes about 30+ seconds of network request to download all the dependencies needed for those specific jobs ([build](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a06f-4f59-bdcb-dbf449560bf4) + [scan](https://scans.gradle.com/s/ffh6h2nccp3ic/performance/network-activity) for `detekt`), which is about the same time it takes for the dependency cache to actually be restored, which is more or less about 20+ seconds ([build](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a40-4ce3-94b0-7834ae800702/195-204)). Thus, this optimization is not really worth it for these specific jobs. For all those other jobs targeted, it takes more than a minute of network requests to download all the dependencies needed ([build](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a07b-4c70-847b-60c599946d6e) + [scan](https://scans.gradle.com/s/cnwge4ncdug2a/performance/network-activity)), and as such the diff is worth it.

PS.2: Although the `Android Test` are commented-out, I chose to include cache restoration into those as well, just in case they will be reactivated at some point in the future.

PS.3: Dependency cache isn't enabled on the [release-builds.yml](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.buildkite/release-builds.yml) pipeline just yet, but just temporarily, and until we become comfortable with running this new process on the main [pipeline.yml](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.buildkite/pipeline.yml) pipeline. Having said that, and due to the fact that the [release-builds.yml](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.buildkite/release-builds.yml) pipeline is also utilizing the [lint.sh](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.buildkite/commands/lint.sh) for its [Lint](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.buildkite/release-builds.yml#L24-L48) steps, due to the fact that the [restore-cache.sh](https://github.com/wordpress-mobile/WordPress-Android/pull/21584/files#diff-268c0a56a3349de264049b135c7bb163ca875e54e3e8eea65eac1de32911b8a2R3) is within that `lint.sh` script, restoring cache will partially work on the [release-builds.yml](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.buildkite/release-builds.yml) pipeline as well.

---

### Future Improvements

Me and @wzieba has been discussing on DMs about some possible future improvement, which are including but are not limited to the below list:

1. Only execute the `assemble android test` task, which depends on `assemble` task anyway, and would most likely resolve most of the dependencies, for all other tasks to benefit from. However, this needs further testing to make sure that running other tasks such us `lint` and `assemble unit test` are providing little to no help.
2. Create a custom Gradle task and utilize Gradle's [init scripts](https://docs.gradle.org/current/userguide/init_scripts.html) (`init.d`). Using the init script, go through all available modules for a project, and then, using that custom Gradle task, resolve all dependencies for all tasks that are available per module. We could take some inspiration from plugins such as [this](https://plugins.gradle.org/plugin/org.gradle.github-dependency-graph-gradle-plugin).
3. Instead of scheduling the `dependency cache` build to run weekly ([buildkite-ci#578](https://github.com/Automattic/buildkite-ci/pull/578)), make it run more often, possibly scheduling it to even run daily. However, since dependencies are not being updated too often, and, even when they do on that specific week, it would only be for a few of them. This will indeed cause some extra network requests, but most of the dependencies cached would still benefit from the existing and weekly refreshed dependency cache.
4. Enable dependency cache on the [release-builds.yml](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.buildkite/release-builds.yml) pipeline.

---

## To Test:

<!-- Test instructions per dependency update: https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/docs/test_instructions_per_dependency_update.md -->

This dependency cache solution was thoroughly tested via this #21550 test-only PR, but to help you asserting that this change is indeed making an impact, plus not causing any trouble for the project, try and follow the below:

- `Save Cache`:

  - Triggering [schedules/dependency-cache.yml](https://github.com/wordpress-mobile/WordPress-Android/pull/21584/files#diff-f1eb78f1f1134bf64e6d4a27fd0b816ae7dff3a0da45f4115e8898bc6939642e) manually via [this](https://buildkite.com/automattic/wordpress-android/builds/20849) build:
    - Starts [downloading the dependencies per task](https://buildkite.com/automattic/wordpress-android/builds/20849#01946ac5-da8a-4bc8-b755-6aada34206aa/193).
    - Caches [the dependencies successfully](https://buildkite.com/automattic/wordpress-android/builds/20849#01946ac5-da8a-4bc8-b755-6aada34206aa/448-449).
      - Duration: [1m 15s](https://buildkite.com/automattic/wordpress-android/builds/20849#01946ac5-da8a-4bc8-b755-6aada34206aa/448-459)
      - Size: [1076.74 MB](https://buildkite.com/automattic/wordpress-android/builds/20849#01946ac5-da8a-4bc8-b755-6aada34206aa/448-460)
    - Total time it takes for the `dependency cache` job to complete: [14m 31s](https://buildkite.com/automattic/wordpress-android/builds/20849#01946ac5-da8a-4bc8-b755-6aada34206aa)
  - Sending a Slack notification (`#android-core-notifs`):
    - On success, notifying that: `The dependency cache has been updated.` (p1736959083969609-slack-C0787KRDYDC)
    - On failure, notifying that: `Failure to update the dependency cache.` (not tested, but should behave similarly to ☝️)
- `Restore Cache`:

| Job | BT (Without) | BT (With) | BT (Diff*) | NR (Without) | NR (With) | RC (Without) | RC (With) | Win |
| ---- | ------------ | ---------- | ---- | ------------- | --------- | -------------- | --------- | ---- |
| Jetpack App | [6m 32s](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a07b-4c70-847b-60c599946d6e) | [4m 38s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a40-4ce3-94b0-7834ae800702) | `- 1m 54s` (✅) | [1m 33](https://scans.gradle.com/s/cnwge4ncdug2a/performance/network-activity) | [0s](https://scans.gradle.com/s/ntufqqu2dbkmo/performance/network-activity) | `0s` | [20s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a40-4ce3-94b0-7834ae800702/195-204) | `1m 13s` (✅) |
| WordPress App | [6m 36s](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a07a-4f9f-8a0c-cec33cf35dcc) | [4m 21s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a3e-4619-864e-7cd75e6d21a3) | `- 2m 15s` (✅) | [1m 31s](https://scans.gradle.com/s/jbhh22fr3tk7q/performance/network-activity) | [0s](https://scans.gradle.com/s/vb5x3x5xpglcy/performance/network-activity) | `0s` | [18s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a3e-4619-864e-7cd75e6d21a3/195-204) | `1m 13s` (✅) |
| Jetpack Lint | [6m 19s](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a071-4754-9b54-a9bb50cb1fe6) | [4m 31s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a2a-48a5-87f0-208a6c70b267) | `- 1m 48s` (✅) | [1m 5s](https://scans.gradle.com/s/u3vc4jissl74y/performance/network-activity) | [2s](https://scans.gradle.com/s/d7ee7yhxm6mpy/performance/network-activity) | `0s` | [18s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a2a-48a5-87f0-208a6c70b267/195-204) | `47s` (✅) |
| WordPress Lint | [6m 35s](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a070-4264-b413-48480d2db90c) | [4m 51s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a29-47d3-9865-8a7398078144) | `- 1m 44s` (✅) | [1m 18s](https://scans.gradle.com/s/akuzzoxuhuibq/performance/network-activity) | [2s](https://scans.gradle.com/s/zyzzssvox3eyc/performance/network-activity) | `0s` | [19s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a29-47d3-9865-8a7398078144/195-204) | `59s` (✅) |
| WordPress Unit Tests | [5m 52s](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a074-4fbd-ab95-9c0ce95a3de6) | [4m 18s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a35-4292-9c99-093641443fca) | `- 1m 34s` (✅) | [1m 32s](https://scans.gradle.com/s/za4o7ncnwxyfu/performance/network-activity) | [4s](https://scans.gradle.com/s/4xcswieicsybs/performance/network-activity) | `0s` | [18s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a35-4292-9c99-093641443fca/230-239) | `1m 14s` (✅) |
| Processors Unit Tests | [3m 43s](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a075-4a41-af79-09c8dd87e163) | [2m 6s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a37-4af6-b1ba-45651e33e715) | `- 1m 37s` (✅) | [1m 23s](https://scans.gradle.com/s/qnaft6y474hnu/performance/network-activity) | [2s](https://scans.gradle.com/s/o6qshfwrdtysu/performance/network-activity) | `0s` | [19s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a37-4af6-b1ba-45651e33e715/230-239) | `1m 4s` (✅) |
| Image Editor Unit Tests | [2m 53s](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a076-4f3e-967f-37c20b39a9d0) | [3m 11s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a38-4bd9-a3a6-2553ae30c9c6) | `+ 18s` (❗) | [40s](https://scans.gradle.com/s/a3sjznraaqyym/performance/network-activity) | [1s](https://scans.gradle.com/s/mtcjvr7tzmu42/performance/network-activity) | `0s` | [34s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a38-4bd9-a3a6-2553ae30c9c6/231-240) (❗️) | `6s` (✅) |
| FluxC Unit Tests | [3m 35s](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a077-4348-8c3d-32cee105b5c8) | [2m 31s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a39-4441-917b-84dbc92b88c2) | `- 1m 4s` (✅) | [48s](https://scans.gradle.com/s/zxrw5wcp3kzuq/performance/network-activity) | [1s](https://scans.gradle.com/s/b7ymuasqn7qqe/performance/network-activity) | `0s` | [19s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a39-4441-917b-84dbc92b88c2/230-239) | `29s` (✅) |
| Login Unit Tests | [4m 12s](https://buildkite.com/automattic/wordpress-android/builds/20845#01946a7e-a078-4162-8698-7ab4d34ea019) | [2m 44s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a3a-4861-9138-aac72299a8b2) | `- 1m 28s` (✅) | [1m 40s](https://scans.gradle.com/s/3v2n3cw5vzwno/performance/network-activity) | [1s](https://scans.gradle.com/s/gtnylta26qmqa/performance/network-activity) | `0s` | [22s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a3a-4861-9138-aac72299a8b2/230-239) | `1m 18s` (✅) |

> [!NOTE]
> - `Without`: Without Dependency Cache
> - `With`: With Dependency Cache
> - `BT`: Build Time
> - `NR`: Network Requests
> - `RC`: Restore Cache
> - `Win`: NR (Without) - RC (With)

> [!WARNING]
> (*): The `BT (Diff)` presented above is relative, so don't read too much into it, it might not match with the `WIN`, which is a much more solid number. But still, it is nice to have and compare those two number together. To to be clear about that, although I tried to pick the most representative builds, without and with the dependency cache mechanism in place, there are other factors that can make any build faster, or slower for that matter (like `build cache`, `firebase test lab`, etc), so again, don't read too much into it.
> 
> (**): Both, the `Jetpack` and `WordPress Android Tests` are currently disabled and as such not included in the above `Restore Cache` related table.

> [!CAUTION]
> (❗️) The dependency cache restoration for `Image Editor Unit Tests` job took almost twice as expected [34s](https://buildkite.com/automattic/wordpress-android/builds/20851#01946ae5-0a38-4bd9-a3a6-2553ae30c9c6/231-240). This is quite interesting and I am not sure how to explain it. For a cache of more-or-less 1 GB in size, and as per all the other jobs suggest, it takes about `20+` seconds to download the cache entry from s3.

---

### Merge Instructions

- [x] Merge this PR.
- [x] Close #21550 and #21540 test-only PRs.

-----

## Regression Notes

1. Potential unintended areas of impact

    - CI builds failing or slower than expected.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To Test` section above.

3. What automated tests I added (or what prevented me from doing so)

    - N/A

---

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

---

## Testing Checklist (strike-out the not-applying and unnecessary ones): `N/A`

ParaskP7 committed:
CI: Update a8c-ci-toolkit to 3.9.1

Release Notes: https://github.com/Automattic/
a8c-ci-toolkit-buildkite-plugin/releases/tag/3.9.1

FYI: This change points to that version of the 'A8C CI Toolkit' where
both, the #135 and #138 PRs, are merged into 'trunk', and, the new
dependency cache mechanism, per project, without 'GRADLE_RO_DEP_CACHE',
is fully operational.

A8C CI Toolkit #135 PR: [Dependency Cache] Dependency Cache on CI per
Project [without GRADLE_RO_DEP_CACHE] #135
- https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/135

A8C CI Toolkit #138 PR: [Dependency Cache] Fix Dependency Cache #138
- https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/138

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 committed:
CI: Add a scheduled dependency cache job (save cache)

FYI: This job will be then used by 'buildkite-ci' and configured as a
'buildkite_pipeline_schedule' with a weekly frequency.

PS.1: The targeted 'pipeline' related jobs are:
- Mobile App
- Lint
- Unit Tests
- Android tests
PS.2: Since the 'WordPress' app is being build from the very same
codebase as the 'Jetpack' app, targeting only the 'Jetpack' related
'pipeline' jobs are enough to download and cache all the dependencies
required for both apps.

ParaskP7 committed:
CI: Fix wrongly defined command related double quotes on pipeline jobs

Otherwise, using 'multi-cmd' will produce 'no such file or directory'
type of errors.

ParaskP7 committed:
CI: Add restore cache and run on targeted pipeline related jobs only

FYI: The targeted 'pipeline' related jobs are:
- Mobile App
- Lint
- Unit Tests
- Android Tests (Commented Out)

PS.1: The 'detekt' and 'checkstyle' jobs aren't targeted because they
only take about 30+ seconds of network request to download all the
dependencies needed for those specific jobs, which is about the same
time it takes for the dependency cache to actually be restored (20+
seconds). Thus, this optimization is not really worth it for these
specific jobs. For all those other jobs targeted, it take more than a
minute of network requests to download all the dependencies needed, and
as such worth the diff is worth it.
PS.2: Although the 'Android Tests' are commented-out, I chose to include
cache restoration into those as well, just in case they will be
reactivated at some point in the future.

ParaskP7 committed:
CI: Move restore cache call to individual .sh script files

WCAndroid PR Comment: https://github.com/woocommerce/
woocommerce-android/pull/13303#discussion_r1916752265

ParaskP7 committed:
CI: Fix wrongly defined command related double quotes on pipeline jobs

When using 'single-cmd' and not using this double quotes structure it is
is producing 'did not find expected key' type of errors.

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into ci/enable-dependency-cache-on-ci-final

ParaskP7 committed:
Docs: Add gradlew test suite to comment on save cache for unit tests

wzieba reviewed (approved):
LGTM!

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into ci/enable-dependency-cache-on-ci-final
